treeherder/perf/investigation_aids.py

```diff
@@ -0,0 +1,216 @@
+            for key, value in source.items():
+                target.setdefault(key, value)
+
+        return target
+
+    @staticmethod
+    def _fetch_job(job_id: str):
+        return (Job.objects
+                .prefetch_related('taskcluster_metadata')
+                .get(id=job_id))
+
+    @classmethod
+    def _filter_relevant_actions(cls, actions_json: dict, original_task) -> List:
+        actions = actions_json['actions']
+
+        def collect_relevant_actions(accumulating_actions, action):
+            acc_action_names = [action['name'] for action in accumulating_actions]
+
+            action_of_known_kind = cls._has_known_kind(action)
+            action_is_relevant = action['name'] not in acc_action_names
+            no_context_or_task_to_check = (not len(action['context'])) and (not original_task)
+            task_is_in_context = (original_task and original_task.get('tags') and
+                                  cls._task_in_context(action['context'], original_task['tags']))
+
+            if (action_of_known_kind and action_is_relevant and
+                    (no_context_or_task_to_check or task_is_in_context)):
+                accumulating_actions.append(action)
+            return accumulating_actions
+
+        actions = reduce(collect_relevant_actions, actions, [])
+        return list(actions)
+
+    @classmethod
+    def _has_known_kind(cls, action):
+        return action['kind'] in cls.KNOWN_KINDS
+
+    @classmethod
+    def _task_in_context(cls, tag_set_list, task_tags):
+        return any(
+            all(task_tags.get(tag) and task_tags[tag] == tag_set[tag]
+                for tag in tag_set.keys())
+            for tag_set in tag_set_list
+        )
+
+
+class BackfillTool:
+
+    def __init__(self, taskcluster_model: TaskclusterModel):
+        self.tc_model = taskcluster_model
+
+    def backfill_job(self, job_id: str) -> str:
+        job = self._fetch_job(job_id)
+
+        if not self.can_backfill(job):
+            logger.warning(f"Cannot backfill job {job_id}.")
+            return ''
+
+        decision_job = self._fetch_associated_decision_job(job)
+        decision_task_id = decision_job.taskcluster_metadata.task_id
+
+        results = self.tc_model.load(decision_task_id, job)
+        backfill_task = self._get_action(results["actions"], "backfill")
+
+        logger.debug(f"Requesting backfill for task {results['originalTaskId']}...")
+        task_id = self.tc_model.submit(
+            action=backfill_task,
+            decision_task_id=decision_task_id,
+            task_id=results["originalTaskId"],
+            input={},
+            static_action_variables=results["staticActionVariables"]
+        )
+        return task_id
+
+    def can_backfill(self, job: Job) -> bool:
+        has_credentials = self.tc_model.has_credentials()
+        is_try_repo = job.repository.is_try_repo
+
+        if not has_credentials:
+            logger.debug("Credentials for backfilling are missing.")
```

Comment:
Not sure this check is needed?

Reply:
@tomprince are you referring to the extra debug logs? I can easily remove them.

Reply:
I think checking whether the repo is try makes sense here, because not backfilling on try is a policy.

I'm not sure checking if credentials are provided here make sense. For example, this could be running via taskcluster-proxy, in which case things will work without credentials.